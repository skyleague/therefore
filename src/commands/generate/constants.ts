import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
