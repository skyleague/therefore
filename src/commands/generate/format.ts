import path from 'node:path'
import { Biome, Distribution } from '@biomejs/js-api'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

const biome = await (async () => {
    const local = await Biome.create({
        distribution: Distribution.NODE,
    })
    try {
        const baseConfig = (await import(`file:///${process.cwd().replace(/\\/g, '/')}/biome.json`)).default
        if ('extends' in baseConfig && Array.isArray(baseConfig.extends)) {
            for (const extend of baseConfig.extends) {
                const extendedConfig = (await import(extend)).default
                local.applyConfiguration(extendedConfig)
            }
        }
        local.applyConfiguration(baseConfig)
    } catch (_e) {
        console.debug('No biome.json found')
    }

    return local
})()

export async function formatFile({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    const formatted = biome.formatContent(input, {
        // this virual path might ignore the file alltogether
        // filePath: file,
        filePath: `src/${path.dirname(file)}.${file.split('.').pop()}`,
    })
    return formatted.content
}
