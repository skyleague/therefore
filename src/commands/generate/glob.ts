import { Nothing, fstat, isJust, isRight, normalizePath } from '@skyleague/axioms'
import fastGlob from 'fast-glob'

import path from 'node:path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.allSettled(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && isJust(eitherStat.right)) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(normalizePath(pattern))
                    }
                    if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(normalizePath(path.relative(cwd, absolutePath)))}/**/*${extension}`
                    }

                    return Nothing
                }
                if (pattern.startsWith('!')) {
                    globOptions.ignore.push(normalizePath(pattern.slice(1)))
                    return Nothing
                }
                return normalizePath(pattern)
            }),
        )
    )
        .map((result) => (result.status === 'fulfilled' ? result.value : Nothing))
        .filter(isJust)
    return await fastGlob(entries, globOptions)
}
