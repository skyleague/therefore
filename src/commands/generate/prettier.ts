import type { ThereforeOutputType } from './types.js'

import { Nothing, isNothing } from '@skyleague/axioms'

export async function maybeLoadPrettier() {
    try {
        return await import('prettier')
    } catch {
        return Nothing
    }
}

export async function formatFile(
    prettier: Awaited<ReturnType<typeof maybeLoadPrettier>>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (isNothing(prettier)) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
