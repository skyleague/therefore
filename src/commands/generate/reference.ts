import type { FileDefinition, TypescriptDefinition, TypescriptReference } from './types'

import type { ThereforeCst } from '../../lib/primitives/types'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript'

import { evaluate, sha256 } from '@skyleague/axioms'

import path from 'path'

interface ExpandedReference {
    srcPath: string
    targetPath: string
    symbolName: string
    uuid: string
    referenceName: string
    uniqueSymbolName: string
}

const localVersion = new Map<string, number>()
const referenceRegister = new Map<string, string>()

export function requireReference({
    definitions,
    current,
    ref,
    locals,
}: {
    definitions: Record<string, FileDefinition>
    current: FileDefinition
    ref: TypescriptReference
    locals: NonNullable<TypescriptDefinition['locals']>
}): ExpandedReference[] {
    const found = Object.values(definitions)
        .flatMap((d) =>
            d.symbols.map((s) => ({
                uuid: s.definition.uuid,
                symbolName: s.definition.symbolName,
                referenceName: s.definition.referenceName,
                uniqueSymbolName: s.definition.uniqueSymbolName,
                srcPath: d.srcPath,
                targetPath: d.targetPath,
            }))
        )
        .find((s) => s.uuid === ref.uuid)

    if (found === undefined) {
        // this is a locally defined variable
        let sourceSymbol: string
        const reference = evaluate(ref.reference[0])

        const refName = ref.name
        const fileHash = sha256(path.basename(current.srcPath)).slice(0, 4)
        if (refName !== undefined) {
            sourceSymbol = refName
        } else {
            if (!referenceRegister.has(reference.uuid)) {
                if (!localVersion.has(current.targetPath)) {
                    localVersion.set(current.targetPath, 0)
                }
                const localIterator = localVersion.get(current.targetPath)!
                referenceRegister.set(reference.uuid, `local${localIterator}`)
                localVersion.set(current.targetPath, localIterator + 1)
            }
            sourceSymbol = referenceRegister.get(reference.uuid)!
        }

        const { definition } = toTypescriptDefinition({
            sourceSymbol,
            schema: evaluate(reference) as ThereforeCst,
            fileHash,
            exportSymbol: ref.exportSymbol,
            locals,
        })
        current.symbols.push({
            uuid: ref.uuid,
            definition,
            symbolName: sourceSymbol,
            typeOnly: true,
        })

        // make the DFS well defined
        const extra = definition.references.map((r) => requireReference({ definitions, current, ref: r, locals })).flat()
        return [
            ...extra,
            {
                uuid: ref.uuid,
                symbolName: definition.symbolName,
                srcPath: current.srcPath,
                targetPath: current.targetPath,
                referenceName: definition.referenceName,
                uniqueSymbolName: definition.uniqueSymbolName,
            },
        ]
    }
    return [
        {
            uuid: ref.uuid,
            symbolName: found.symbolName,
            srcPath: found.srcPath,
            targetPath: found.targetPath,
            referenceName: found.referenceName,
            uniqueSymbolName: found.uniqueSymbolName,
        },
    ]
}
