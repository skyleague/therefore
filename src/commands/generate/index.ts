import { generate } from './generate.js'

import { constants } from '../../lib/constants.js'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((p) => p.toString()),
        extension: constants.schemaExtension,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
