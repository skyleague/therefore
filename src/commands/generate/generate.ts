import { generatedAjv, generatedBy } from './constants.js'
import { expandGlobs } from './glob.js'
import { formatFile, maybeLoadPrettier } from './prettier.js'
import { resolveTypescriptSchema } from './resolver.js'
import type { FileDefinition, OutputFile, ReferenceData, ThereforeOutputType } from './types.js'

import { getExtension, replaceExtension } from '../../common/template/path.js'
import { renderTemplate } from '../../common/template/template.js'
import type { ThereforeNode } from '../../lib/cst/cst.js'
import type { ThereforeCst } from '../../lib/primitives/types.js'
import { isThereforeExport } from '../../lib/primitives/types.js'
import { toJsonSchema } from '../../lib/visitor/jsonschema/jsonschema.js'
import { prepass } from '../../lib/visitor/prepass/prepass.js'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript.js'

import { entriesOf, enumerate, groupBy, isDefined, range, second } from '@skyleague/axioms'
import decamelize from 'decamelize'

import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, ThereforeNode | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/').replace(/\.ts$/, '.js')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, ThereforeNode | unknown>
}

export async function loadSymbol({
    srcPath,
    basePath,
    symbol,
    sourceSymbol,
    compile,
    definitions,
    entry,
    fileSuffix,
    filePath,
    outputFileRename,
    appendSourceSymbolName: appendSourceName = false,
    rootPath = srcPath,
}: {
    srcPath: string
    basePath: string
    symbol: ThereforeCst
    sourceSymbol: string
    compile: boolean
    definitions: Record<string, FileDefinition | undefined>
    entry: string
    fileSuffix?: string | undefined
    filePath?: string | undefined
    outputFileRename: (path: string) => string
    appendSourceSymbolName?: boolean
    rootPath?: string
}) {
    const simplified = prepass(symbol)
    let targetDir = '.'
    if (simplified.type === 'custom') {
        if (simplified.value.fileSuffix !== undefined) {
            fileSuffix = simplified.value.fileSuffix
        }
        if (simplified.value.filePath !== undefined) {
            const dir = path.dirname(simplified.value.filePath)
            const basename = path.basename(simplified.value.filePath)
            filePath = basename
            targetDir = dir
        }
    }

    const { base: baseName, dir: baseDir } = path.parse(srcPath)
    const ext = getExtension(srcPath) ?? ''
    const targetBaseDir = path.relative(basePath, path.join(baseDir, targetDir))
    const targetSrcPath = `${targetBaseDir}/${(filePath ?? baseName).replace(ext, '')}${ext}`
    const targetPath = fileSuffix !== undefined ? replaceExtension(targetSrcPath, fileSuffix) : outputFileRename(targetSrcPath)

    definitions[targetPath] ??= {
        srcPath: path.relative(basePath, rootPath),
        targetPath,
        attachedFiles: [],
        symbols: [],
        dependencies: {},
    }

    const file = definitions[targetPath]!
    // check if symbol is not seen yet
    if (file.symbols.find((s) => s.uuid === simplified.uuid)) {
        return
    }
    const symbolName = simplified.name ?? sourceSymbol

    sourceSymbol = appendSourceName && simplified.name !== undefined ? `${sourceSymbol}.${simplified.name}` : symbolName
    const schemaName = decamelize(symbolName, { separator: '-' })
    const schemaFile = `./schemas/${schemaName}.schema.json`
    const compiledFile = `./schemas/${schemaName}.schema.js`

    const nodeIsCompiled = simplified.description.validator?.compile ?? compile

    const jsonschema = await toJsonSchema(simplified, nodeIsCompiled)
    const { definition, subtrees } = toTypescriptDefinition({ sourceSymbol, symbolName, schema: simplified })

    file.symbols.push(
        ...Object.values(definition.locals ?? {})
            .filter(isDefined)
            .map((local) => ({
                uuid: local.uuid,
                symbolName: local.symbolName,
                definition: local,
                typeOnly: true,
            }))
    )

    file.symbols.push({
        uuid: simplified.uuid,
        symbolName,
        definition: definition,
        schemaFile,
        compiledFile: nodeIsCompiled ? compiledFile : undefined,
        typeOnly: simplified.description.validator?.enabled !== true,
    })

    if (definition.isExported && simplified.description.validator?.enabled) {
        const cleanSchemasFolder = () => {
            const targetFolder = path.dirname(path.join(targetBaseDir, compiledFile))
            if (fs.existsSync(targetFolder)) {
                console.warn(`Cleaning ${targetFolder}`)
                fs.rmSync(targetFolder, { force: true, recursive: true })
            }
        }
        if (jsonschema.compiled) {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, compiledFile),
                content: `/**\n * ${generatedAjv} \n * eslint-disable\n */\n${jsonschema.code}`,
                prettify: false,
                type: 'validator',
                clean: cleanSchemasFolder,
            })
        } else {
            file.attachedFiles.push({
                targetPath: path.join(targetBaseDir, schemaFile),
                content: JSON.stringify(jsonschema.schema, null, 2),
                prettify: true,
                type: 'jsonschema',
                clean: cleanSchemasFolder,
            })
        }

        console.debug(` - found ${definition.symbolName}`)
    }
    for (const { node: subSymbol, fileSuffix: subFileSuffix, filePath: subFilePath } of subtrees) {
        await loadSymbol({
            symbol: subSymbol,
            basePath,
            sourceSymbol,
            compile,
            definitions,
            srcPath: path.join(basePath, targetPath),
            rootPath: srcPath,
            fileSuffix: subFileSuffix,
            filePath: subFilePath,
            entry,
            outputFileRename,
            appendSourceSymbolName: true,
        })
    }
}

export async function scanModule({
    entry,
    srcPath,
    basePath,
    compile,
    definitions,
    require = requireModule,
    outputFileRename,
}: {
    entry: string
    srcPath: string
    basePath: string
    compile: boolean
    definitions: Record<string, FileDefinition>
    require?: (module: string) => Promise<Record<string, ThereforeNode | unknown>>
    outputFileRename: (path: string) => string
}) {
    const module = await require(entry)

    for (const [sourceSymbol, symbolPromise] of Object.entries(module)) {
        const symbol = await symbolPromise

        if (!isThereforeExport(symbol)) {
            continue
        }

        await loadSymbol({ symbol, sourceSymbol, compile, definitions, srcPath, entry, basePath, outputFileRename })
    }
}

export async function scanFiles({
    files,
    basePath,
    compile,
    outputFileRename,
}: {
    files: string[]
    basePath: string
    compile: boolean
    outputFileRename: (path: string) => string
}): Promise<Record<string, FileDefinition>> {
    const definitions: Record<string, FileDefinition> = {}
    for (const entry of files) {
        const srcPath = path.resolve(basePath, entry)
        let isGenerated = false
        const strippedPath = srcPath.substr(0, srcPath.lastIndexOf('.'))
        for (const ext of ['.ts', '.js']) {
            const p = `${strippedPath}${ext}`
            isGenerated ||= fs.existsSync(p) && fs.readFileSync(p).includes(generatedBy)
        }
        if (isGenerated) {
            console.debug(`scanning ${entry}`)
            console.debug(` * skipping generated schema`)
            continue
        }

        console.log(`scanning ${entry}`)
        try {
            await scanModule({ entry, basePath, srcPath, compile, definitions, outputFileRename })
        } catch (e: unknown) {
            const error = e as Error
            console.debug(error.message, error.stack)
            throw error
        }
    }
    return definitions
}

export async function compileOutputFiles(
    entries: string[],
    { compile = false, cwd, outputFileRename }: { cwd: string; compile: boolean; outputFileRename: (path: string) => string }
): Promise<
    {
        targetPath: string
        template: string
        data: Record<string, ReferenceData>
        type: ThereforeOutputType
        prettify: boolean
        clean?: () => void
    }[]
> {
    const outputFiles: OutputFile[] = []
    const definitions = await scanFiles({ files: entries, basePath: cwd, compile, outputFileRename })

    const localReferences = {}

    for (const targetPath of Object.keys(definitions)) {
        resolveTypescriptSchema({ definitions, targetPath, cwd, outputFiles, localReferences })
    }

    return compile ? outputFiles : outputFiles.filter((f) => f.type !== 'validator')
}

export async function generate({
    globs,
    ignore,
    extension,
    compile,
    outputFileRename,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    compile: boolean
    outputFileRename: (path: string) => string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const outputFiles = await compileOutputFiles(entries, {
        cwd,
        outputFileRename,
        compile,
    })

    if (clean) {
        for (const of of outputFiles) {
            if (of.clean !== undefined) {
                of.clean()
            }
        }
    }

    const prettier = maybeLoadPrettier()

    const data = outputFiles.reduce<Record<string, ReferenceData>>(
        (agg, x) => ({
            ...agg,
            ...x.data,
        }),
        {}
    )

    const duplicates = entriesOf(groupBy(entriesOf(data), ([, values]) => `${values.srcPath}${values.symbolName}`)).filter(
        ([, records]) => records.length > 1
    )
    for (const duplicateGroup of duplicates.map((d) => second(d))) {
        for (const [i, [key]] of enumerate(duplicateGroup)) {
            // data[key].referenceName = `${data[key].referenceName}${i}`
            data[key]!.symbolName = `${data[key]!.symbolName}${i}`
            data[key]!.uniqueSymbolName = `${data[key]!.uniqueSymbolName}${i}`
        }
    }

    const variablesMap = Object.fromEntries(
        Object.entries(data)
            .map(([key, vals]) => Object.entries(vals).map(([varName, value]) => [`${key}:${varName}`, value] as const))
            .flat()
    )

    for (const { targetPath, template, type, prettify } of outputFiles) {
        fs.mkdirSync(path.dirname(targetPath), { recursive: true })
        // allow nested references up to 3 deep
        let contents = template
        for (const _ of range(3)) {
            contents = renderTemplate(contents, variablesMap)
        }
        fs.writeFileSync(targetPath, prettify ? await formatFile(prettier, contents, targetPath, type) : contents)
    }
}
