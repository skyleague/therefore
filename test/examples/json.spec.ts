import { compileOutput } from '../../src/commands/generate/generate.js'

import { expect, it } from 'vitest'

it('json', async () => {
    expect(
        await compileOutput(['examples/json/json.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})

it('simple', async () => {
    expect(
        await compileOutput(['examples/json/simple.schema.ts'], {
            cwd: process.cwd(),
        }),
    ).toMatchSnapshot()
})
