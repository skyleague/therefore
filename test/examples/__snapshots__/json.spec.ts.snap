// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/json/json.type.ts": "/**
 * Generated by @skyleague/therefore@v1.0.0-local
 * Do not manually touch this
 */
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as JsonAdvValidator } from './schemas/json-adv.schema.js'
import { validate as JsonValidator } from './schemas/json.schema.js'

export type Json =
    | string
    | null
    | boolean
    | number
    | {
          [k: string]: Json | undefined
      }
    | Json[]

export const Json = {
    validate: JsonValidator as ValidateFunction<Json>,
    get schema() {
        return Json.validate.schema
    },
    get errors() {
        return Json.validate.errors ?? undefined
    },
    is: (o: unknown): o is Json => Json.validate(o) === true,
    parse: (o: unknown): { right: Json } | { left: DefinedError[] } => {
        if (Json.is(o)) {
            return { right: o }
        }
        return { left: (Json.errors ?? []) as DefinedError[] }
    },
} as const

export type JsonAdv = JsonLocal

export const JsonAdv = {
    validate: JsonAdvValidator as ValidateFunction<JsonAdv>,
    get schema() {
        return JsonAdv.validate.schema
    },
    get errors() {
        return JsonAdv.validate.errors ?? undefined
    },
    is: (o: unknown): o is JsonAdv => JsonAdv.validate(o) === true,
    parse: (o: unknown): { right: JsonAdv } | { left: DefinedError[] } => {
        if (JsonAdv.is(o)) {
            return { right: o }
        }
        return { left: (JsonAdv.errors ?? []) as DefinedError[] }
    },
} as const

type JsonLocal =
    | string
    | null
    | boolean
    | number
    | {
          [k: string]: JsonAdv | undefined
      }
    | JsonAdv[]
",
    "examples/json/schemas/json-adv.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"JsonAdv","$ref":"#/$defs/JsonLocal","$defs":{"JsonLocal":{"anyOf":[{"type":"string"},{"type":"null"},{"type":"boolean"},{"type":"number"},{"type":"object","additionalProperties":{"$ref":"#"}},{"type":"array","items":{"$ref":"#"}}]}}};const schema12 = {"anyOf":[{"type":"string"},{"type":"null"},{"type":"boolean"},{"type":"number"},{"type":"object","additionalProperties":{"$ref":"#"}},{"type":"array","items":{"$ref":"#"}}]};const wrapper0 = {validate: validate10};function validate11(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;const _errs1 = errors;if(typeof data !== "string"){const err0 = {instancePath,schemaPath:"#/anyOf/0/type",keyword:"type",params:{type: "string"},message:"must be string"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}var _valid0 = _errs1 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs3 = errors;if(data !== null){const err1 = {instancePath,schemaPath:"#/anyOf/1/type",keyword:"type",params:{type: "null"},message:"must be null"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}var _valid0 = _errs3 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs5 = errors;if(typeof data !== "boolean"){const err2 = {instancePath,schemaPath:"#/anyOf/2/type",keyword:"type",params:{type: "boolean"},message:"must be boolean"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;}var _valid0 = _errs5 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs7 = errors;if(!((typeof data == "number") && (isFinite(data)))){const err3 = {instancePath,schemaPath:"#/anyOf/3/type",keyword:"type",params:{type: "number"},message:"must be number"};if(vErrors === null){vErrors = [err3];}else {vErrors.push(err3);}errors++;}var _valid0 = _errs7 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs9 = errors;if(errors === _errs9){if(data && typeof data == "object" && !Array.isArray(data)){for(const key0 in data){const _errs12 = errors;if(!(wrapper0.validate(data[key0], {instancePath:instancePath+"/" + key0.replace(/~/g, "~0").replace(/\\//g, "~1"),parentData:data,parentDataProperty:key0,rootData}))){vErrors = vErrors === null ? wrapper0.validate.errors : vErrors.concat(wrapper0.validate.errors);errors = vErrors.length;}var valid1 = _errs12 === errors;if(!valid1){break;}}}else {const err4 = {instancePath,schemaPath:"#/anyOf/4/type",keyword:"type",params:{type: "object"},message:"must be object"};if(vErrors === null){vErrors = [err4];}else {vErrors.push(err4);}errors++;}}var _valid0 = _errs9 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs13 = errors;if(errors === _errs13){if(Array.isArray(data)){var valid2 = true;const len0 = data.length;for(let i0=0; i0<len0; i0++){const _errs15 = errors;if(!(wrapper0.validate(data[i0], {instancePath:instancePath+"/" + i0,parentData:data,parentDataProperty:i0,rootData}))){vErrors = vErrors === null ? wrapper0.validate.errors : vErrors.concat(wrapper0.validate.errors);errors = vErrors.length;}var valid2 = _errs15 === errors;if(!valid2){break;}}}else {const err5 = {instancePath,schemaPath:"#/anyOf/5/type",keyword:"type",params:{type: "array"},message:"must be array"};if(vErrors === null){vErrors = [err5];}else {vErrors.push(err5);}errors++;}}var _valid0 = _errs13 === errors;valid0 = valid0 || _valid0;}}}}}if(!valid0){const err6 = {instancePath,schemaPath:"#/anyOf",keyword:"anyOf",params:{},message:"must match a schema in anyOf"};if(vErrors === null){vErrors = [err6];}else {vErrors.push(err6);}errors++;validate11.errors = vErrors;return false;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}validate11.errors = vErrors;return errors === 0;}function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;if(!(validate11(data, {instancePath,parentData,parentDataProperty,rootData}))){vErrors = vErrors === null ? validate11.errors : vErrors.concat(validate11.errors);errors = vErrors.length;}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/json/schemas/json.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Json","anyOf":[{"type":"string"},{"type":"null"},{"type":"boolean"},{"type":"number"},{"type":"object","additionalProperties":{"$ref":"#"}},{"type":"array","items":{"$ref":"#"}}]};const wrapper0 = {validate: validate10};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;const _errs1 = errors;if(typeof data !== "string"){const err0 = {instancePath,schemaPath:"#/anyOf/0/type",keyword:"type",params:{type: "string"},message:"must be string"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}var _valid0 = _errs1 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs3 = errors;if(data !== null){const err1 = {instancePath,schemaPath:"#/anyOf/1/type",keyword:"type",params:{type: "null"},message:"must be null"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}var _valid0 = _errs3 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs5 = errors;if(typeof data !== "boolean"){const err2 = {instancePath,schemaPath:"#/anyOf/2/type",keyword:"type",params:{type: "boolean"},message:"must be boolean"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;}var _valid0 = _errs5 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs7 = errors;if(!((typeof data == "number") && (isFinite(data)))){const err3 = {instancePath,schemaPath:"#/anyOf/3/type",keyword:"type",params:{type: "number"},message:"must be number"};if(vErrors === null){vErrors = [err3];}else {vErrors.push(err3);}errors++;}var _valid0 = _errs7 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs9 = errors;if(errors === _errs9){if(data && typeof data == "object" && !Array.isArray(data)){for(const key0 in data){const _errs12 = errors;if(!(wrapper0.validate(data[key0], {instancePath:instancePath+"/" + key0.replace(/~/g, "~0").replace(/\\//g, "~1"),parentData:data,parentDataProperty:key0,rootData}))){vErrors = vErrors === null ? wrapper0.validate.errors : vErrors.concat(wrapper0.validate.errors);errors = vErrors.length;}var valid1 = _errs12 === errors;if(!valid1){break;}}}else {const err4 = {instancePath,schemaPath:"#/anyOf/4/type",keyword:"type",params:{type: "object"},message:"must be object"};if(vErrors === null){vErrors = [err4];}else {vErrors.push(err4);}errors++;}}var _valid0 = _errs9 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs13 = errors;if(errors === _errs13){if(Array.isArray(data)){var valid2 = true;const len0 = data.length;for(let i0=0; i0<len0; i0++){const _errs15 = errors;if(!(wrapper0.validate(data[i0], {instancePath:instancePath+"/" + i0,parentData:data,parentDataProperty:i0,rootData}))){vErrors = vErrors === null ? wrapper0.validate.errors : vErrors.concat(wrapper0.validate.errors);errors = vErrors.length;}var valid2 = _errs15 === errors;if(!valid2){break;}}}else {const err5 = {instancePath,schemaPath:"#/anyOf/5/type",keyword:"type",params:{type: "array"},message:"must be array"};if(vErrors === null){vErrors = [err5];}else {vErrors.push(err5);}errors++;}}var _valid0 = _errs13 === errors;valid0 = valid0 || _valid0;}}}}}if(!valid0){const err6 = {instancePath,schemaPath:"#/anyOf",keyword:"anyOf",params:{},message:"must match a schema in anyOf"};if(vErrors === null){vErrors = [err6];}else {vErrors.push(err6);}errors++;validate10.errors = vErrors;return false;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
  },
}
`;

exports[`simple 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/json/schemas/simple.schema.js": "/* eslint-disable */
// @ts-nocheck
/**
 * Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code
 */
"use strict";
/** @type {unknown} */
export const validate = validate10;export default validate10;const schema11 = {"$schema":"http://json-schema.org/draft-07/schema#","title":"Simple","anyOf":[{"type":"number"},{"type":"array","items":{"$ref":"#"}}]};const wrapper0 = {validate: validate10};function validate10(data, {instancePath="", parentData, parentDataProperty, rootData=data}={}){let vErrors = null;let errors = 0;const _errs0 = errors;let valid0 = false;const _errs1 = errors;if(!((typeof data == "number") && (isFinite(data)))){const err0 = {instancePath,schemaPath:"#/anyOf/0/type",keyword:"type",params:{type: "number"},message:"must be number"};if(vErrors === null){vErrors = [err0];}else {vErrors.push(err0);}errors++;}var _valid0 = _errs1 === errors;valid0 = valid0 || _valid0;if(!valid0){const _errs3 = errors;if(errors === _errs3){if(Array.isArray(data)){var valid1 = true;const len0 = data.length;for(let i0=0; i0<len0; i0++){const _errs5 = errors;if(!(wrapper0.validate(data[i0], {instancePath:instancePath+"/" + i0,parentData:data,parentDataProperty:i0,rootData}))){vErrors = vErrors === null ? wrapper0.validate.errors : vErrors.concat(wrapper0.validate.errors);errors = vErrors.length;}var valid1 = _errs5 === errors;if(!valid1){break;}}}else {const err1 = {instancePath,schemaPath:"#/anyOf/1/type",keyword:"type",params:{type: "array"},message:"must be array"};if(vErrors === null){vErrors = [err1];}else {vErrors.push(err1);}errors++;}}var _valid0 = _errs3 === errors;valid0 = valid0 || _valid0;}if(!valid0){const err2 = {instancePath,schemaPath:"#/anyOf",keyword:"anyOf",params:{},message:"must match a schema in anyOf"};if(vErrors === null){vErrors = [err2];}else {vErrors.push(err2);}errors++;validate10.errors = vErrors;return false;}else {errors = _errs0;if(vErrors !== null){if(_errs0){vErrors.length = _errs0;}else {vErrors = null;}}}validate10.errors = vErrors;return errors === 0;};validate.schema=schema11;",
    "examples/json/simple.type.ts": "/**
 * Generated by @skyleague/therefore@v1.0.0-local
 * Do not manually touch this
 */
/* eslint-disable */

import type { DefinedError, ValidateFunction } from 'ajv'

import { validate as SimpleValidator } from './schemas/simple.schema.js'

export type Simple = number | Simple[]

export const Simple = {
    validate: SimpleValidator as ValidateFunction<Simple>,
    get schema() {
        return Simple.validate.schema
    },
    get errors() {
        return Simple.validate.errors ?? undefined
    },
    is: (o: unknown): o is Simple => Simple.validate(o) === true,
    parse: (o: unknown): { right: Simple } | { left: DefinedError[] } => {
        if (Simple.is(o)) {
            return { right: o }
        }
        return { left: (Simple.errors ?? []) as DefinedError[] }
    },
} as const
",
  },
}
`;
